syntax = "proto3";

service MessageService {
  rpc RaftRequestVote(RaftRequestVoteRequest) returns (RaftRequestVoteResponse) {}
  rpc RaftUpdateState(RaftUpdateStateRequest) returns (EmptyResponse) {}

  rpc Authenticate(AuthenticateRequest) returns (StringResponse) {}
  rpc List(ListRequest) returns (StringResponse) {}
  rpc Send(SendRequest) returns (EmptyResponse) {}
  rpc Deliver(DeliverRequest) returns (StringResponse) {}
  rpc Delete(DeleteRequest) returns (EmptyResponse) {}
}

message RaftRequestVoteRequest {
  int32 term = 1;
  string candidate_id = 2;
}

message RaftRequestVoteResponse {
  bool vote = 1;
}

message RaftUpdateStateRequest {
  string replica_id = 1;
  bytes data = 2;
}

message AuthenticateRequest {
  string username = 1;
  string password = 2;
}

message ListRequest {
  string token = 1;
  string request = 2;
}

message SendRequest {
  string token = 1;
  string username = 2;
  string body = 3;
}

message DeliverRequest {
  string token = 1;
}

message DeleteRequest {
  string token = 1;
  string username = 2;
}

message StringResponse {
  bool success = 1;
  string response = 2;
}

message EmptyResponse {
  bool success = 1;
}
